server/src/test/java/org/elasticsearch/env/NodeEnvironmentTests.java

```diff
@@ -523,6 +538,66 @@ public void testBlocksDowngradeToVersionWithMultipleNodesInDataPath() throws IOE
         }
     }
 
+    public void testIndexCompatibilityChecks() throws IOException {
+        final Settings settings = buildEnvSettings(Settings.EMPTY);
+
+        try (NodeEnvironment env = newNodeEnvironment(settings)) {
+
+            try (
+                PersistedClusterStateService.Writer writer = new PersistedClusterStateService(
+                    env.nodeDataPaths(),
+                    env.nodeId(),
+                    xContentRegistry(),
+                    new ClusterSettings(settings, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS),
+                    () -> 0L
+                ).createWriter()
+            ) {
+                writer.writeFullStateAndCommit(
+                    1L,
+                    ClusterState.builder(ClusterName.DEFAULT)
+                        .metadata(
+                            Metadata.builder()
+                                .persistentSettings(Settings.builder().put(Metadata.SETTING_READ_ONLY_SETTING.getKey(), true).build())
+                                .build()
+                        )
+                        .build()
+                );
+            }
+
+            Version oldIndexVersion = Version.fromId(between(1, Version.CURRENT.minimumIndexCompatibilityVersion().id - 1));
+            overrideOldestIndexVersion(oldIndexVersion, env.nodeDataPaths());
+
+            IllegalStateException ex = expectThrows(
+                IllegalStateException.class,
+                "Must fail the check on index that's too old",
+                () -> checkForIndexCompatibility(logger, env.nodePaths())
+            );
+
+            assertThat(ex.getMessage(), containsString("[" + oldIndexVersion + "] exist"));
+            assertThat(ex.getMessage(), startsWith("cannot upgrade node because incompatible indices created with version"));
+
+            // This should work
+            overrideOldestIndexVersion(Version.CURRENT.minimumIndexCompatibilityVersion(), env.nodeDataPaths());
+            checkForIndexCompatibility(logger, env.nodePaths());
+
+            // Trying to boot with newer version should pass this check
+            overrideOldestIndexVersion(NodeMetadataTests.tooNewVersion(), env.nodeDataPaths());
+            checkForIndexCompatibility(logger, env.nodePaths());
+
+            // Simulate empty old index version, attempting to upgrade before 7.17
+            removeOldestIndexVersion(env.nodeDataPaths());
+
+            ex = expectThrows(
+                IllegalStateException.class,
+                "Must fail the check on index that's too old",
+                () -> checkForIndexCompatibility(logger, env.nodePaths())
+            );
+
+            assertThat(ex.getMessage(), startsWith("cannot upgrade node because incompatible indices created with version"));
```

Comment:
Aha yeah this is what I meant in a previous comment - we should report this as the node version being too old rather than mentioning `V_EMPTY`.